### ISSUE TITLE
MYSQL (no viable alternative at input)

### ISSUE THREAD

**Issue by kuangyoubo** [apache/shardingsphere#5452]
## Bug Report

**For English only**, other languages will not accept.

Before report a bug, make sure you have:

- Searched open and closed [GitHub issues](https://github.com/apache/shardingsphere/issues).
- Read documentation: [ShardingSphere Doc](https://shardingsphere.apache.org/document/current/en/overview).

Please pay attention on issues you submitted, because we maybe need more details. 
If no response **more than 7 days** and we cannot reproduce it on current information, we will **close it**.

Please answer these questions before submitting your issue. Thanks!

### Which version of ShardingSphere did you use?

4.1.0

### Which project did you use? Sharding-JDBC or Sharding-Proxy?
Sharding-JDBC

### Expected behavior
No output log

### Actual behavior
line 1:45 no viable alternative at input 'CONVERT(substring(customer_code,3),signed'
line 1:45 no viable alternative at input 'CONVERT(substring(customer_code,3),signed'

### Reason analyze (If you can)
Unsupported function

### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.

select CONVERT(substring(customer_code, 3) , signed) as maxCustNo from tb_sp_user_customer order by id desc limit 1

### Example codes for reproduce this issue (such as a github link).
select CONVERT(substring(customer_code, 3) , signed) as maxCustNo from tb_sp_user_customer order by id desc limit 1

**Comment by kimmking** [comment]
CONVERT is not supported now.
You can use
```
select cast(m as signed) as maxCustNo from (select substring(customer_code,3) as m from tb_sp_user_customer order by id desc limit 1) t;
```


In my case:
```
mysql-proxy> select cast(m as signed) as maxCustNo from (select substring(name, -1) as m from aa order by id desc limit 1) t;
+-----------+
| maxCustNo |
+-----------+
|         8 |
+-----------+
1 row in set (0.02 sec)
```

**Comment by kuangyoubo** [comment]
> CONVERT is not supported now.
> You can use
> 
> ```
> select cast(m as signed) as maxCustNo from (select substring(customer_code,3) as m from tb_sp_user_customer order by id desc limit 1) t;
> ```
> 
> In my case:
> 
> ```
> mysql-proxy> select cast(m as signed) as maxCustNo from (select substring(name, -1) as m from aa order by id desc limit 1) t;
> +-----------+
> | maxCustNo |
> +-----------+
> |         8 |
> +-----------+
> 1 row in set (0.02 sec)
> ```

Can't nest functions ?


> ## Bug Report
> **For English only**, other languages will not accept.
> 
> Before report a bug, make sure you have:
> 
> * Searched open and closed [GitHub issues](https://github.com/apache/shardingsphere/issues).
> * Read documentation: [ShardingSphere Doc](https://shardingsphere.apache.org/document/current/en/overview).
> 
> Please pay attention on issues you submitted, because we maybe need more details.
> If no response **more than 7 days** and we cannot reproduce it on current information, we will **close it**.
> 
> Please answer these questions before submitting your issue. Thanks!
> 
> ### Which version of ShardingSphere did you use?
> 4.1.0
> 
> ### Which project did you use? Sharding-JDBC or Sharding-Proxy?
> Sharding-JDBC
> 
> ### Expected behavior
> No output log
> 
> ### Actual behavior
> line 1:45 no viable alternative at input 'CONVERT(substring(customer_code,3),signed'
> line 1:45 no viable alternative at input 'CONVERT(substring(customer_code,3),signed'
> 
> ### Reason analyze (If you can)
> Unsupported function
> 
> ### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.
> select CONVERT(substring(customer_code, 3) , signed) as maxCustNo from tb_sp_user_customer order by id desc limit 1
> 
> ### Example codes for reproduce this issue (such as a github link).
> select CONVERT(substring(customer_code, 3) , signed) as maxCustNo from tb_sp_user_customer order by id desc limit 1



> CONVERT is not supported now.
> You can use
> 
> ```
> select cast(m as signed) as maxCustNo from (select substring(customer_code,3) as m from tb_sp_user_customer order by id desc limit 1) t;
> ```
> 
> In my case:
> 
> ```
> mysql-proxy> select cast(m as signed) as maxCustNo from (select substring(name, -1) as m from aa order by id desc limit 1) t;
> +-----------+
> | maxCustNo |
> +-----------+
> |         8 |
> +-----------+
> 1 row in set (0.02 sec)
> ```

Can't nest functions?

**Comment by kuangyoubo** [comment]
> CONVERT is not supported now.
> You can use
> 
> ```
> select cast(m as signed) as maxCustNo from (select substring(customer_code,3) as m from tb_sp_user_customer order by id desc limit 1) t;
> ```
> 
> In my case:
> 
> ```
> mysql-proxy> select cast(m as signed) as maxCustNo from (select substring(name, -1) as m from aa order by id desc limit 1) t;
> +-----------+
> | maxCustNo |
> +-----------+
> |         8 |
> +-----------+
> 1 row in set (0.02 sec)
> ```

Can't nest functions?

**Comment by kuangyoubo** [comment]
> CONVERT is not supported now.
> You can use
> 
> ```
> select cast(m as signed) as maxCustNo from (select substring(customer_code,3) as m from tb_sp_user_customer order by id desc limit 1) t;
> ```
> 
> In my case:
> 
> ```
> mysql-proxy> select cast(m as signed) as maxCustNo from (select substring(name, -1) as m from aa order by id desc limit 1) t;
> +-----------+
> | maxCustNo |
> +-----------+
> |         8 |
> +-----------+
> 1 row in set (0.02 sec)
> ```

select cast(m as signed) as maxUserCode from (select substring(user_code,3) as m from tb_sp_user order by id desc limit 1) t


line 1:18 no viable alternative at input 'cast(massigned'

**Comment by kimmking** [comment]
Oh...
I try the master version 5.0...

**Comment by kuangyoubo** [comment]
> Oh...
> I try the master version 5.0...
> 4.x is not supporting subquery.

When will shardingsphere 5.0 be released?

**Comment by kuangyoubo** [comment]
Now many of the MYSQL syntax are not supported, such as recursion, subquery.


In my case:

with recursive r as (select * from tb_sp_user where user_code = 'BS100000168'  
 union all select c.* from tb_sp_user c,r where c.COM_CODE=r.user_code and c.STORE_FLAG='1' 
 and c.register_type not in ('8','9'))select user_code,user_name,com_code,MOBILE from r ;

**Comment by tuohai666** [comment]
@kuangyoubo It seems like a parser problem, I'll investigate it.

**Comment by herbertsgit** [comment]
I have a similar problem
Version: ShardingSphere-JDBC  4.1.0

The SQL I executed : UPDATE table  SET REQUEST_ID = ?,PROGRAM_ID = ?,OBJECT_VERSION_NUMBER=OBJECT_VERSION_NUMBER+1,LAST_UPDATED_BY = ?,LAST_UPDATE_DATE=CURRENT_TIMESTAMP,WHERE  LINE_ID = ?  AND OBJECT_VERSION_NUMBER=?

Error:  no viable alternative at input 'CURRENT_TIMESTAMP,'

Another SQL : SELECT
	s.FUNCTION_ID AS id,
	s.FUNCTION_CODE AS CODE,
	t.FUNCTION_NAME AS text,
	s.FUNCTION_ICON AS icon,
	s.PARENT_FUNCTION_ID AS parentId,
	t.lang 
FROM
	sys_function_b s
	LEFT JOIN sys_function_tl t ON s.FUNCTION_ID = t.FUNCTION_ID 
WHERE
	t.lang IS NOT NULL

Error: no viable alternative at input 'text'
mismatched input 'text' expecting {TRUNCATE, POSITION, VIEW, ANY, OFFSET, BEGIN, COMMIT, ROLLBACK, SAVEPOINT, BOOLEAN, DATE, TIME, TIMESTAMP, YEAR, QUARTER, MONTH, WEEK, DAY, HOUR, MINUTE, SECOND, MICROSECOND, MAX, MIN, SUM, COUNT, AVG, CURRENT, ENABLE, DISABLE, INSTANCE, DO, DEFINER, CASCADED, LOCAL, CLOSE, OPEN, NEXT, NAME, TYPE, TABLES, TABLESPACE, COLUMNS, FIELDS, INDEXES, STATUS, MODIFY, VALUE, DUPLICATE, FIRST, LAST, AFTER, OJ, ACCOUNT, USER, ROLE, START, TRANSACTION, WITHOUT, ESCAPE, SUBPARTITION, STORAGE, SUPER, TEMPORARY, THAN, UNBOUNDED, SIGNED, UPGRADE, VALIDATION, ROLLUP, SOUNDS, UNKNOWN, OFF, ALWAYS, COMMITTED, LEVEL, NO, PASSWORD, PRIVILEGES, ACTION, ALGORITHM, AUTOCOMMIT, BTREE, CHAIN, CHARSET, CHECKSUM, CIPHER, CLIENT, COALESCE, COMMENT, COMPACT, COMPRESSED, COMPRESSION, CONNECTION, CONSISTENT, DATA, DISCARD, DISK, ENCRYPTION, END, ENGINE, EVENT, EXCHANGE, EXECUTE, FILE, FIXED, FOLLOWING, GLOBAL, HASH, IMPORT_, LESS, MEMORY, NONE, PARSER, PARTIAL, PARTITIONING, PERSIST, PRECEDING, PROCESS, PROXY, QUICK, REBUILD, REDUNDANT, RELOAD, REMOVE, REORGANIZE, REPAIR, REVERSE, SESSION, SHUTDOWN, SIMPLE, SLAVE, VISIBLE, INVISIBLE, ENFORCED, AGAINST, LANGUAGE, MODE, QUERY, EXTENDED, EXPANSION, VARIANCE, MAX_ROWS, MIN_ROWS, SQL_BIG_RESULT, SQL_BUFFER_RESULT, SQL_CACHE, SQL_NO_CACHE, STATS_AUTO_RECALC, STATS_PERSISTENT, STATS_SAMPLE_PAGES, ROW_FORMAT, WEIGHT_STRING, COLUMN_FORMAT, INSERT_METHOD, KEY_BLOCK_SIZE, PACK_KEYS, PERSIST_ONLY, BIT_AND, BIT_OR, BIT_XOR, GROUP_CONCAT, JSON_ARRAYAGG, JSON_OBJECTAGG, STD, STDDEV, STDDEV_POP, STDDEV_SAMP, VAR_POP, VAR_SAMP, AUTO_INCREMENT, AVG_ROW_LENGTH, DELAY_KEY_WRITE, ROTATE, MASTER, BINLOG, ERROR, SCHEDULE, COMPLETION, EVERY, HOST, SOCKET, PORT, SERVER, WRAPPER, OPTIONS, OWNER, RETURNS, CONTAINS, SECURITY, INVOKER, TEMPTABLE, MERGE, UNDEFINED, DATAFILE, FILE_BLOCK_SIZE, EXTENT_SIZE, INITIAL_SIZE, AUTOEXTEND_SIZE, MAX_SIZE, NODEGROUP, WAIT, LOGFILE, UNDOFILE, UNDO_BUFFER_SIZE, REDO_BUFFER_SIZE, HANDLER, PREV, ORGANIZATION, DEFINITION, DESCRIPTION, REFERENCE, FOLLOWS, PRECEDES, IMPORT, CONCURRENT, XML, DUMPFILE, SHARE, CODE, CONTEXT, SOURCE, CHANNEL, CLONE, AGGREGATE, INSTALL, COMPONENT, UNINSTALL, RESOURCE, EXPIRE, NEVER, HISTORY, OPTIONAL, REUSE, MAX_QUERIES_PER_HOUR, MAX_UPDATES_PER_HOUR, MAX_CONNECTIONS_PER_HOUR, MAX_USER_CONNECTIONS, RETAIN, RANDOM, OLD, ISSUER, SUBJECT, CACHE, GENERAL, SLOW, USER_RESOURCES, EXPORT, RELAY, HOSTS, FLUSH, RESET, RESTART, IO_THREAD, SQL_THREAD, SQL_BEFORE_GTIDS, SQL_AFTER_GTIDS, MASTER_LOG_FILE, MASTER_LOG_POS, RELAY_LOG_FILE, RELAY_LOG_POS, SQL_AFTER_MTS_GAPS, UNTIL, DEFAULT_AUTH, PLUGIN_DIR, STOP, IDENTIFIER_, STRING_}

**Comment by herbertsgit** [comment]
> @kuangyoubo It seems like a parser problem, I'll investigate it.

I have a similar problem ,how can i do with this @tuohai666 
Version: ShardingSphere-JDBC 4.1.0

The SQL I executed : UPDATE table SET REQUEST_ID = ?,PROGRAM_ID = ?,OBJECT_VERSION_NUMBER=OBJECT_VERSION_NUMBER+1,LAST_UPDATED_BY = ?,LAST_UPDATE_DATE=CURRENT_TIMESTAMP,WHERE LINE_ID = ? AND OBJECT_VERSION_NUMBER=?

Error: no viable alternative at input 'CURRENT_TIMESTAMP,'

Another SQL : SELECT
s.FUNCTION_ID AS id,
s.FUNCTION_CODE AS CODE,
t.FUNCTION_NAME AS text,
s.FUNCTION_ICON AS icon,
s.PARENT_FUNCTION_ID AS parentId,
t.lang
FROM
sys_function_b s
LEFT JOIN sys_function_tl t ON s.FUNCTION_ID = t.FUNCTION_ID
WHERE
t.lang IS NOT NULL

Error: no viable alternative at input 'text'
mismatched input 'text' expecting {TRUNCATE, POSITION, VIEW, ANY, OFFSET, BEGIN, COMMIT, ROLLBACK, SAVEPOINT, BOOLEAN, DATE, TIME, TIMESTAMP, YEAR, QUARTER, MONTH, WEEK, DAY, HOUR, MINUTE, SECOND, MICROSECOND, MAX, MIN, SUM, COUNT, AVG, CURRENT, ENABLE, DISABLE, INSTANCE, DO, DEFINER, CASCADED, LOCAL, CLOSE, OPEN, NEXT, NAME, TYPE, TABLES, TABLESPACE, COLUMNS, FIELDS, INDEXES, STATUS, MODIFY, VALUE, DUPLICATE, FIRST, LAST, AFTER, OJ, ACCOUNT, USER, ROLE, START, TRANSACTION, WITHOUT, ESCAPE, SUBPARTITION, STORAGE, SUPER, TEMPORARY, THAN, UNBOUNDED, SIGNED, UPGRADE, VALIDATION, ROLLUP, SOUNDS, UNKNOWN, OFF, ALWAYS, COMMITTED, LEVEL, NO, PASSWORD, PRIVILEGES, ACTION, ALGORITHM, AUTOCOMMIT, BTREE, CHAIN, CHARSET, CHECKSUM, CIPHER, CLIENT, COALESCE, COMMENT, COMPACT, COMPRESSED, COMPRESSION, CONNECTION, CONSISTENT, DATA, DISCARD, DISK, ENCRYPTION, END, ENGINE, EVENT, EXCHANGE, EXECUTE, FILE, FIXED, FOLLOWING, GLOBAL, HASH, IMPORT_, LESS, MEMORY, NONE, PARSER, PARTIAL, PARTITIONING, PERSIST, PRECEDING, PROCESS, PROXY, QUICK, REBUILD, REDUNDANT, RELOAD, REMOVE, REORGANIZE, REPAIR, REVERSE, SESSION, SHUTDOWN, SIMPLE, SLAVE, VISIBLE, INVISIBLE, ENFORCED, AGAINST, LANGUAGE, MODE, QUERY, EXTENDED, EXPANSION, VARIANCE, MAX_ROWS, MIN_ROWS, SQL_BIG_RESULT, SQL_BUFFER_RESULT, SQL_CACHE, SQL_NO_CACHE, STATS_AUTO_RECALC, STATS_PERSISTENT, STATS_SAMPLE_PAGES, ROW_FORMAT, WEIGHT_STRING, COLUMN_FORMAT, INSERT_METHOD, KEY_BLOCK_SIZE, PACK_KEYS, PERSIST_ONLY, BIT_AND, BIT_OR, BIT_XOR, GROUP_CONCAT, JSON_ARRAYAGG, JSON_OBJECTAGG, STD, STDDEV, STDDEV_POP, STDDEV_SAMP, VAR_POP, VAR_SAMP, AUTO_INCREMENT, AVG_ROW_LENGTH, DELAY_KEY_WRITE, ROTATE, MASTER, BINLOG, ERROR, SCHEDULE, COMPLETION, EVERY, HOST, SOCKET, PORT, SERVER, WRAPPER, OPTIONS, OWNER, RETURNS, CONTAINS, SECURITY, INVOKER, TEMPTABLE, MERGE, UNDEFINED, DATAFILE, FILE_BLOCK_SIZE, EXTENT_SIZE, INITIAL_SIZE, AUTOEXTEND_SIZE, MAX_SIZE, NODEGROUP, WAIT, LOGFILE, UNDOFILE, UNDO_BUFFER_SIZE, REDO_BUFFER_SIZE, HANDLER, PREV, ORGANIZATION, DEFINITION, DESCRIPTION, REFERENCE, FOLLOWS, PRECEDES, IMPORT, CONCURRENT, XML, DUMPFILE, SHARE, CODE, CONTEXT, SOURCE, CHANNEL, CLONE, AGGREGATE, INSTALL, COMPONENT, UNINSTALL, RESOURCE, EXPIRE, NEVER, HISTORY, OPTIONAL, REUSE, MAX_QUERIES_PER_HOUR, MAX_UPDATES_PER_HOUR, MAX_CONNECTIONS_PER_HOUR, MAX_USER_CONNECTIONS, RETAIN, RANDOM, OLD, ISSUER, SUBJECT, CACHE, GENERAL, SLOW, USER_RESOURCES, EXPORT, RELAY, HOSTS, FLUSH, RESET, RESTART, IO_THREAD, SQL_THREAD, SQL_BEFORE_GTIDS, SQL_AFTER_GTIDS, MASTER_LOG_FILE, MASTER_LOG_POS, RELAY_LOG_FILE, RELAY_LOG_POS, SQL_AFTER_MTS_GAPS, UNTIL, DEFAULT_AUTH, PLUGIN_DIR, STOP, IDENTIFIER_, STRING_}

**Comment by tuohai666** [comment]
@kuangyoubo Thanks for your report. 
Can you provide more information about the expect behavior and actual behavior?
I think you may want to describe what's you expect on the MySQL CLI/JDBC, and the actual thing you see from MySQL CLI/JDBC.

Please provide the imformation straightforwardly and originally without any process or analyze.

**Comment by tuohai666** [comment]
@herbertsgit Thanks for your report.
Can you create another issue to track that problem? Because it's not a similiar problem.

**Comment by tuohai666** [comment]
Any way, I have fixed this problem, for 5.0.0 and 4.1.1(release soon).

**Comment by JOYS369** [comment]
> Anyway, I have fixed this problem, for 5.0.0 and 4.1.1(will be released soon).

4.1.1,my project can't strat.
the error message is :mismatched input 'text' expecting {TRUNCATE,...,IDENTIFIER_, STRING_}
